dockstore-webservice/src/main/java/io/dockstore/webservice/languages/IpynbHandler.java

```diff
@@ -0,0 +1,247 @@
+        Optional<SourceFile> file = sourceFiles.stream().filter((sourceFile -> Objects.equals(sourceFile.getAbsolutePath(), notebookPath))).findFirst();
+        if (!file.isPresent()) {
+            return negativeValidation(notebookPath, "No notebook file is present");
+        }
+        String content = file.get().getContent();
+
+        // Parse the notebook JSON.
+        JSONObject notebook;
+        try {
+            notebook = new JSONObject(content);
+        } catch (JSONException ex) {
+            return negativeValidation(notebookPath, "The notebook file is not valid JSON", ex);
+        }
+
+        // Confirm the existence and typedness of the fields that should always be present.
+        try {
+            checkEssentialFields(notebook);
+        } catch (JSONException ex) {
+            return negativeValidation(notebookPath, "The notebook file is malformed", ex);
+        }
+
+        // Check that the entry's programming language (descriptor language subclass) matches the notebook's programming language.
+        try {
+            String entryLanguage = workflow.getDescriptorTypeSubclass().toString();
+            String notebookLanguage = extractProgrammingLanguage(notebook);
+            if (!entryLanguage.equalsIgnoreCase(notebookLanguage)) {
+                return negativeValidation(notebookPath, String.format("The notebook programming language must be '%s'", entryLanguage));
+            }
+        } catch (JSONException ex) {
+            return negativeValidation(notebookPath, "Error reading the notebook programming language", ex);
+        }
+
+        // If we reach this point, everything is well, return a positive validation.
+        return positiveValidation();
+    }
+
+    private void checkEssentialFields(JSONObject notebook) throws JSONException {
+        // The JSONObject.get* methods throw if the key is missing or not set to the expected type of value.
+        notebook.getJSONObject(METADATA_KEY);
+        notebook.getInt(NBFORMAT_KEY);
+        notebook.getInt(NBFORMAT_MINOR_KEY);
+        notebook.getJSONArray(CELLS_KEY);
+    }
+
+    private String extractProgrammingLanguage(JSONObject notebook) throws JSONException {
+        JSONObject metadata = notebook.getJSONObject(METADATA_KEY);
+        // If key "language_info" is present, the spec says that it must contain the key "name".
+        if (metadata.has(LANGUAGE_INFO_KEY)) {
+            return metadata.getJSONObject(LANGUAGE_INFO_KEY).optString(NAME_KEY, PYTHON);
+        }
+        return PYTHON;
+    }
+
+    @Override
+    public VersionTypeValidation validateToolSet(Set<SourceFile> sourceFiles, String primaryDescriptorFilePath) {
+        throw new UnsupportedOperationException("Notebooks do not support tools");
+    }
+
+    @Override
+    public VersionTypeValidation validateTestParameterSet(Set<SourceFile> sourceFiles) {
+        throw new UnsupportedOperationException("Notebooks do not support test files");
```

Comment:
Should probably rephrase to something more like  "Dockstore does not support notebook test files yet."
I assume they probably have something like https://www.wrighters.io/parameters-jupyter-notebooks-with-papermill/